clients/pkg/promtail/discovery/dockerd/dockerd.go

```diff
@@ -0,0 +1,320 @@
+	tg := &targetgroup.Group{
+		Source: "Docker",
+	}
+
+	containers, err := d.client.ContainerList(ctx, types.ContainerListOptions{Filters: d.filters})
+	if err != nil {
+		return nil, fmt.Errorf("error while listing containers: %w", err)
+	}
+
+	networkLabels, err := getNetworksLabels(ctx, d.client, dockerLabel)
+	if err != nil {
+		return nil, fmt.Errorf("error while computing network labels: %w", err)
+	}
+
+	for _, c := range containers {
+		commonLabels := map[string]string{
+			dockerLabelContainerID:          c.ID,
+			dockerLabelContainerName:        c.Names[0],
+			dockerLabelContainerNetworkMode: c.HostConfig.NetworkMode,
+		}
+
+		for k, v := range c.Labels {
+			ln := strutil.SanitizeLabelName(k)
+			commonLabels[dockerLabelContainerLabelPrefix+ln] = v
+		}
+		containerAdd := false
+		for _, n := range c.NetworkSettings.Networks {
+			var added bool
+			for _, p := range c.Ports {
+				if p.Type != "tcp" {
+					continue
+				}
+
+				labels := model.LabelSet{
+					dockerLabelNetworkIP:   model.LabelValue(n.IPAddress),
+					dockerLabelPortPrivate: model.LabelValue(strconv.FormatUint(uint64(p.PrivatePort), 10)),
+				}
+
+				if p.PublicPort > 0 {
+					labels[dockerLabelPortPublic] = model.LabelValue(strconv.FormatUint(uint64(p.PublicPort), 10))
+					labels[dockerLabelPortPublicIP] = model.LabelValue(p.IP)
+				}
+
+				for k, v := range commonLabels {
+					labels[model.LabelName(k)] = model.LabelValue(v)
+				}
+
+				for k, v := range networkLabels[n.NetworkID] {
+					labels[model.LabelName(k)] = model.LabelValue(v)
+				}
+
+				addr := net.JoinHostPort(n.IPAddress, strconv.FormatUint(uint64(p.PrivatePort), 10))
+				labels[model.AddressLabel] = model.LabelValue(addr)
+				tg.Targets = append(tg.Targets, labels)
+				added = true
+			}
+
+			if !added {
+				// Use fallback port when no exposed ports are available or if all are non-TCP
+				labels := model.LabelSet{
+					dockerLabelNetworkIP: model.LabelValue(n.IPAddress),
+				}
+
+				for k, v := range commonLabels {
+					labels[model.LabelName(k)] = model.LabelValue(v)
+				}
+
+				for k, v := range networkLabels[n.NetworkID] {
+					labels[model.LabelName(k)] = model.LabelValue(v)
+				}
+
+				// Containers in host networking mode don't have ports,
+				// so they only end up here, not in the previous loop.
+				var addr string
+				if c.HostConfig.NetworkMode != "host" {
+					addr = net.JoinHostPort(n.IPAddress, strconv.FormatUint(uint64(d.port), 10))
+				} else {
+					addr = d.hostNetworkingHost
+				}
+
+				labels[model.AddressLabel] = model.LabelValue(addr)
+				tg.Targets = append(tg.Targets, labels)
+			}
+			containerAdd = true
+		}
+		if !containerAdd {
+			labels := model.LabelSet{}
+			for k, v := range commonLabels {
+				labels[model.LabelName(k)] = model.LabelValue(v)
+			}
+			addr := d.hostNetworkingHost
+			labels[model.AddressLabel] = model.LabelValue(addr)
+			tg.Targets = append(tg.Targets, labels)
+
+		}
```

Comment:
According to a diff with [docker.go](https://github.com/prometheus/prometheus/blob/main/discovery/moby/docker.go) from the Prometheus discovery only this is new. I'm sorry, I don't really understand how this is solving the problem for you.

Reply:
> According to a diff with [docker.go](https://github.com/prometheus/prometheus/blob/main/discovery/moby/docker.go) from the Prometheus discovery only this is new. I'm sorry, I don't really understand how this is solving the problem for you.

the different place:
<img width="1399" alt="image" src="https://user-images.githubusercontent.com/57797847/221451685-2d7b80b9-7886-4fe0-8831-e1c7a5f02aa8.png">